onnx/defs/function.cc

```diff
@@ -60,9 +60,20 @@ void FunctionExpandHelper(
 
   // For undefined attributes of the function node
   // add default values obtained from the function schema.
+  // get the domain verison for function schema
+  int domain_version = -1;
+  for(const auto& opset_import : func.opset_import()){
+    if(opset_import.domain() == node.domain()) {
+      domain_version = static_cast<int>(opset_import.version());
+    }
+  }
+  if(domain_version == -1) {
+    ONNX_THROW("No opset import registered for domain '" + node.domain() + "' in function proto");
```

Comment:
IIUC, the check here means that every function in a domain `"foo"` must have an entry for `"foo"` in its `opset_import` with a particular version, which becomes `domain_version` here. Is that right?

This logic seems a little tortuous... would it make more sense to have a `domain_version` attribute on the `FunctionProto`? Or - perhaps better - the version could be inferred from the `opset_import` on the containing model (or inferred from the op schema, if the function is not a model-local function)?

At the very least, we need to document this requirement, because it was not obvious to me from reading the protobuf spec.

[EDIT: I see code in schema.cc that appears to add this information if it's missing. Is it still required for model-local functions?]

Reply:
Right every function in a domain "Foo" must have an entry for "foo" in its opset_import with a particular version.

For function ops - If this info is missing then it is inferred ( from op schema ) and implicitly added during schema finalization.

For model local functions - this info is required to be present. We can infer the version for the function from model opset imports but I wanted to reduce the differences in construction or interpretation of FunctionProto for FunctionOps vs model local functions so I decided not to make an exception for model local functions in this case. 

Let me know your thoughts...

Reply:
There seem to be some corner cases where it is worth documenting what the standard requires. (a) If we have a model-local function "bar" in a domain "foo", is the model/graph required to have an import for "foo" (even though the model does not contain any calls to anything other than "foo")? (b) If so, if the FunctionProto for "bar" also includes an import for "foo", is it allowed to be different from what the graph specifies? The opset-compatibility discussion does not clarify this special-case. It doesn't seem important to allow them to be different in this case. (c) If we don't allow to be different, forcing the model to include the same information in two places seems less desirable. So, inferring it from the model's imports may be reasonable.

Reply:
In general, it seems that model-local functions should not need to use opset-imports and should be able to get away with the model's own opset-imports (for the ops called from the function body). So, doing the same for the function's own domain (look it up from the model's imports) seems reasonable. It is also important to clarify which of the three options we go with (i) required, (ii) allowed but not required, or (iii) not allowed. In functions outside models, we do need to allow opset imports.